src/apps/lwaftr/ipv6_apps.lua

```diff
@@ -186,8 +185,24 @@ function NDP:push()
    local isouth, osouth = self.input.south, self.output.south
    local inorth, onorth = self.input.north, self.output.north
    if self.do_ns_request then
-      self.do_ns_request = false -- TODO: have retries, etc
-      transmit(osouth, packet.clone(self.ns_pkt))
+      self.do_ns_request = false
+      local ns_pkt = ndp.form_ns(self.conf.src_eth, self.conf.src_ipv6,
+         self.conf.dst_ipv6)
+
+      -- Send new NS solicitation every second.
+      local retry_ns = timer.new("retry_ns", function()
+         transmit(osouth, packet.clone(ns_pkt))
+      end, 1e9, "repeating")
+      timer.activate(retry_ns)
+
+      -- Abort execution if NS_RESOLVE_THRESHOLD expires.
+      local retry_ns_threshold = timer.new("retry_ns_threshold", function()
+         error(("Could not resolve IPv6 address: %s"):format(
+            ipv6:ntop(self.conf.dst_ipv6)))
+         main.exit(1)
+      end, NS_RESOLVE_THRESHOLD * 1e9)
```

Comment:
I'd almost certainly make "retry_ns_threshold" and main.exit(1) start on the same column.

I'd probably put the timer.new on a new line and/or define the function in a local, because it's more than a line long.

I would absolutely definitely not have the 'end, NS_RESOLVE_THESHOLD * 1e9)' line with the same indentation as the local.